zkevm-circuits/src/state_circuit/lexicographic_ordering.rs

```diff
@@ -0,0 +1,298 @@
+}
+
+impl<F: Field> Queries<F> {
+    fn new(meta: &mut VirtualCells<'_, F>, config: Config, rotation: Rotation) -> Self {
+        let mut query_advice = |column| meta.query_advice(column, rotation);
+        Self {
+            // witness: query_advice(config.witness),
+            tag: query_advice(config.tag),
+            field_tag: query_advice(config.field_tag),
+            id_limbs: config.id_limbs.map(&mut query_advice),
+            address_limbs: config.address_limbs.map(&mut query_advice),
+            storage_key_bytes: config.storage_key_bytes.map(&mut query_advice),
+            rw_counter_limbs: config.rw_counter_limbs.map(query_advice),
+        }
+    }
+
+    fn packed_tags(&self) -> Expression<F> {
+        (1u64 << 4).expr() * self.tag.clone() + self.field_tag.clone()
+    }
+
+    fn storage_key_limbs(&self) -> Vec<Expression<F>> {
+        self.storage_key_bytes
+            .iter()
+            .tuples()
+            .map(|(hi, lo)| (1u64 << 16).expr() * hi.clone() + lo.clone())
+            .collect()
+    }
+
+    fn be_limbs(&self) -> Vec<Expression<F>> {
+        let mut limbs: Vec<_> = self
+            .id_limbs
+            .iter()
+            .rev()
+            .chain(self.address_limbs.iter().rev())
+            .chain(self.storage_key_limbs().iter().rev())
+            .chain(self.rw_counter_limbs.iter().rev())
+            .cloned()
+            .collect();
+        // most significant byte of id should be 0, so safe to overwrite it with packed
+        // tags.
+        limbs[0] = limbs[0].clone() + self.packed_tags() * (1u64 << 8).expr();
+        limbs
+    }
+}
+
+fn rw_to_be_limbs(row: &Rw) -> Vec<u16> {
+    let mut be_bytes = vec![];
+    be_bytes.extend_from_slice(&(row.id().unwrap_or_default() as u32).to_be_bytes());
+    assert_eq!(be_bytes.len(), 4);
+
+    // check that the first byte of id is not used, and overwrites it with packed
+    // tags.
+    assert_eq!(be_bytes[0], 0);
+    be_bytes[0] = row.field_tag().unwrap_or_default() as u8 + ((row.tag() as u8) << 4);
+    assert_eq!(be_bytes.len(), 4);
+
+    be_bytes.extend_from_slice(&(row.address().unwrap_or_default().0));
+    assert_eq!(be_bytes.len(), 24);
+    be_bytes.extend_from_slice(&(row.storage_key().unwrap_or_default().to_be_bytes()));
+    assert_eq!(be_bytes.len(), 24 + 32);
+    be_bytes.extend_from_slice(&((row.rw_counter() as u32).to_be_bytes()));
+    assert_eq!(be_bytes.len(), 24 + 32 + 4);
+
+    be_bytes
+        .iter()
+        .tuples()
+        .map(|(hi, lo)| u16::from_be_bytes([*hi, *lo]))
+        .collect()
+}
+
+fn diff_1_possible_values<F: Field>(cur: Queries<F>, prev: Queries<F>) -> Vec<Expression<F>> {
+    let mut result = vec![];
+    let mut partial_sum = 0u64.expr();
+    for (cur_limb, prev_limb) in cur.be_limbs()[..15].iter().zip(&prev.be_limbs()[..15]) {
+        partial_sum = partial_sum * (1u64 << 16).expr() + cur_limb.clone() - prev_limb.clone();
+        result.push(partial_sum.clone())
```

Comment:
here we first get `partial_sum` vector, which may contain polys of degree [1,2,3,4,... ] then out of `diff_1_possible_values`, we use `.fold(1.expr(), Expression::mul)`.  So i am a bit afraid the final degree is O(N**2)?   It will be horrible..   I am not sure, you can print the degree and check it ?

Reply:
each expression in 
```
fn diff_1_possible_values<F: Field>(cur: Queries<F>, prev: Queries<F>) -> Vec<Expression<F>> {
    let mut result = vec![];
    let mut partial_sum = 0u64.expr();
    for (cur_limb, prev_limb) in cur.be_limbs()[..15].iter().zip(&prev.be_limbs()[..15]) {
        partial_sum = partial_sum * (1u64 << 16).expr() + cur_limb.clone() - prev_limb.clone();
        result.push(partial_sum.clone())
    }
    result
}
``` 
is only degree one, since it's just a linear combination of the limbs. I added a degree test which asserts that the state circuit degree is 16, 15 from product + 1 for the selector on the fixed column.

Reply:
yes you are right